### PR TITLE
Flattens the policy document objects to accept pure strings

### PR DESCRIPTION
**Pull request by lorengordon**
This will require that users separately resolve the policy_document,
any way they like. They can use the included submodule, `policy_document`,
which accepts the same `template`, `template_paths`, and `template_vars`
arguments that were used previously.

This pull request's main purpose is to update the `policy_document` module
to support `any` as the template_var type, which is supported by the
`templatefile()` function. However, there is a limitation in terraform's
collections, while you can have a `list(any)` for example, and it will indeed
accept any type, all objects in the list still must be the same type. Kind of
defeats the purpose IMO, but it is what it is.

The prior approach ran afoul of this limitation, when the user provided
`template_vars` per policy_document, and different `template_vars` objects
had different types. Which is not unlikely because different policy_documents
would map to different policies, and have different policy template
variables. Some policy templates might even use jsonencode with complex
objects, which works fine with `templatefile()`. But this fails when providing
lists of `template_vars` where each item may have different types.

So instead, we're now asking the user to resolve the policy_documents
separately. This allows them to resolve complex types in the templates,
in any way they like that will actually work within terraform's constraints.

### CODE REVIEW

**Comment by lorengordon** [comment]
results of `make test`:

```
PASS
ok      tardigarde-ci/tests     326.036s
[terratest/test]: Completed successfully!
```

File: modules/policy_document/main.tf [hunk 1]
```diff
@@ -19,73 +19,73 @@ locals {
   ]
 }
 
-data aws_iam_policy_document empty {}
+data "aws_iam_policy_document" "empty" {}
```
Comment (kbalk) [inline modules/policy_document/main.tf:22]: Regarding line 12:  How do you ensure that you will have no more than 10 documents?   Or what happens if you have more than 10 documents?  Do they just override the "aws_iam_policy_document ten"?

Comment (lorengordon) [reply modules/policy_document/main.tf:22]: Very good question. It is a limitation of this implementation that `template_paths` only supports up to 10 values. We test for that in the variable definition using the variable validation feature that is new in tf 0.13, here: https://github.com/plus3it/terraform-aws-tardigrade-iam-principals/blob/1c7e59e2a9f5dbb8c4f21a9dfb4bd5459034cf13/modules/policy_document/variables.tf#L6-L13

Comment (kbalk) [reply modules/policy_document/main.tf:22]: Missed that and I started with policy_document.

File: modules/user/main.tf [hunk 2]
```diff
@@ -44,17 +34,17 @@ resource aws_iam_user_policy_attachment this {
 }
 
 # create inline policies for the IAM users
-resource aws_iam_user_policy this {
-  for_each = local.inline_policies
+resource "aws_iam_user_policy" "this" {
+  for_each = { for policy in var.inline_policies : policy.name => policy }
```
Comment (kbalk) [inline modules/user/main.tf:38]: I am missing something very basic here ... how does the policy field within inline_policies get set?  Looking at a tests/create_users_policies/main.tf, I can see it's pulling data from policy_documents.  So policy is a policy document, which is a template?  And where does 'name' come from, or is that one of those variables that is already associated with that object?

Comment (lorengordon) [reply modules/user/main.tf:38]: Right, so the loop here is iterating over `var.inline_policies`, which has this definition:

https://github.com/lorengordon/terraform-aws-tardigrade-iam-principals/blob/29918a1b4218b84cc13504247f48649809a0a1b0/modules/user/variables.tf#L28-L35

The user must set all the attributes of that object, which includes both `name` and `policy`. In the construct in this for expression, there are two "policy" items... the policy object being iterated over (e.g. `for *policy* in ...`), and the policy string that is an [attribute of each policy object](https://github.com/lorengordon/terraform-aws-tardigrade-iam-principals/blob/29918a1b4218b84cc13504247f48649809a0a1b0/modules/user/variables.tf#L32) (which would be referenced as `policy.policy` if we needed it within this loop).

At this point, the `policy.policy` string is a fully resolved policy document. It is not a template. (That's a fundamental change in this PR. Only the policy_document module accepts policy _templates_. The policy reference in every other module is now a resolved policy document.)

In the tests, we define the `inline_policies` in a local, where you can see `name` and `policy` being set for each inline_policy. The user chooses and sets the name, it is of some significance to them, and this module uses the inline policy name as the resource label for the for_each expression.

https://github.com/lorengordon/terraform-aws-tardigrade-iam-principals/blob/29918a1b4218b84cc13504247f48649809a0a1b0/tests/create_users/main.tf#L76-L85

There were a few questions here, not sure I fully answered them all, but hopefully I got far enough that the pieces make more sense?

Comment (kbalk) [reply modules/user/main.tf:38]: Re: var.inline_policies ... my bad .. the policy field is pretty obvious; that's the second time I missed something in the variables.tf files.  Re: policy.policy - I was looking at this snippet

https://github.com/lorengordon/terraform-aws-tardigrade-iam-principals/blob/29918a1b4218b84cc13504247f48649809a0a1b0/modules/policy_document/main.tf#L5-L20

where it's setting a local policy_documents list using the values in the template_paths, template_vars and template variables.  So we're going from the local policy_documents list in modules.policy_document to the variable inline_policies in modules.user, which essentially means the template path is used for the inline_policies 'name' field and the template_vars is used for the inline_policies 'policy' field?

Comment (lorengordon) [reply modules/user/main.tf:38]: In this implementation, the `name` fields are all set by the user, and referenced by the user. All explicitly. No implicit linkages buried in the code. So the name of the policy and the name of the policy_document do not _have_ to be the same. But in the policy object, the reference to the policy document does need to use the name specified in the policy document. Saying "policy" a lot, so that could be hard to follow.

The "create_all" test demonstrates this... Here is a policy_document for an inline policy, the `name` is `"tardigrade-alpha-inline-${local.test_id}"`:

https://github.com/lorengordon/terraform-aws-tardigrade-iam-principals/blob/29918a1b4218b84cc13504247f48649809a0a1b0/tests/create_all/main.tf#L176-L179

And here is the inline_policy that uses that policy_document:

https://github.com/lorengordon/terraform-aws-tardigrade-iam-principals/blob/29918a1b4218b84cc13504247f48649809a0a1b0/tests/create_all/main.tf#L204-L207

The name of the inline_policy is `"tardigrade-alpha-${local.test_id}"`, but its policy object references into the policy_documents module to retrieve the policy with the name `"tardigrade-alpha-inline-${local.test_id}"`.

Comment (kbalk) [reply modules/user/main.tf:38]: Ok.  I can see what you're saying about the name difference.  I'm a bit lost though in general and I'm concerned about the time I'm taking to look at the code.  I missing some very high level concepts here.

From a high level, let's say a caller provides user template info through the objects described in the top level README.  main.tf has a module "users" that sources modules/user.  modules/user/variables.tf has variables for the fields in that object and modules/user/main.tf creates resources with various combos of those variables.  But I don't understand how those variables are used to populate the policy_documents module.

Comment (lorengordon) [reply modules/user/main.tf:38]: These are all good questions. :) The top-level module does not provide resources to resolve the policy template. We _do_ provide the submodule `policy_document` to give the user a mechanism to resolve a policy template and generate the policy document. Users need to invoke that module separately. Here is the example from the create_all test:

https://github.com/lorengordon/terraform-aws-tardigrade-iam-principals/blob/29918a1b4218b84cc13504247f48649809a0a1b0/tests/create_all/main.tf#L15-L22

We invoke the `policy_document` module directly, it resolves the templates to policy documents, and we reference the resolved policy documents in the variables passed to the top-level module.

Comment (kbalk) [reply modules/user/main.tf:38]: The point about policy_document being separate was not clear to me and I must have missed it in the discussions.  It explains a lot.  So that means that while callers/invokers are limited to 10 policy_documents, there are no programmatic limits to the number of individual (group, role, user, policy) objects? 

Shouldn't the README.md for policy_documents specify the limit of 10?

I don't have any other comments on the code itself, but I haven't gone through the test cases.  I think I saw one instance in a test case where the user 'object' didn't initialize or mention the 'depends_on_policies' field/variable, but I suppose there could be a case of no dependence?

Comment (lorengordon) [reply modules/user/main.tf:38]: The `template_paths` argument to each policy_document is limited to 10 paths. The caller can use for_each when invoking the policy_document module to create multiple policy_documents, each one with a different value for its `template_paths` argument, if necessary. There is no limit on the number of policy documents created in this manner.

I can update the template_paths description to mention the max of 10 items...

The `depends_on_policies` is something of an internal variable... It is used by the top-level module to ensure ordering of resource creation and avoid errors of the type "value not known until apply". It's a bit of a wart, to be honest, but I can't think of another way around it without adding a lot more complexity.

**Comment by kbalk** [comment]
This is just a question ... in the snippet below, why did you use local variables, whereas previously you used the data variables:

https://github.com/lorengordon/terraform-aws-tardigrade-iam-principals/blob/flatten-policies/tests/create_all/main.tf#L249-L253

I am done with the review and do not have anything to report, but I am not knowledgable about Terraform and easily could have missed something.  I did not review the create_group*, nor the create_policies tests, but I think I've spent enough time on the differences in this pull request.  It was a bit harder doing the review of the test cases as the code was shuffled around a bit making it appear there were more differences than in actual fact.  I'm not complaining about the shuffle as I can see how you were trying to organize the code; it just took more work to find the real differences.

**Comment by lorengordon** [comment]
>This is just a question ... in the snippet below, why did you use local variables, whereas previously you used the data variables:
>
>https://github.com/lorengordon/terraform-aws-tardigrade-iam-principals/blob/flatten-policies/tests/create_all/main.tf#L249-L253

Mostly just because the "data" references are really long, and I needed the values in more than one place. So I used the locals to collect the values in one place, then used the local references everywhere else.

File: tests/create_policies/main.tf [hunk 3]
```diff
@@ -1,39 +1,53 @@
-provider aws {
+provider "aws" {
   region = "us-east-1"
 }
 
-module policies {
-  source   = "../..//modules/policy"
+module "policies" {
+  source   = "../../modules/policy"
```
Comment (eemperor) [inline tests/create_policies/main.tf:6]: You switch to a single `/` for the relative path in some places but not others.  Any reason why?

File: tests/create_roles/main.tf [hunk 4]
```diff
@@ -1,9 +1,9 @@
-provider aws {
+provider "aws" {
   region = "us-east-1"
 }
 
-module create_roles {
-  source   = "../..//modules/role"
+module "create_roles" {
+  source   = "../../modules/role"
```
Comment (eemperor) [inline tests/create_roles/main.tf:6]: Same as previous comment.

File: tests/create_roles/prereq/main.tf [hunk 5]
```diff
@@ -1,23 +1,56 @@
-provider aws {
+provider "aws" {
   region = "us-east-1"
 }
 
-module policies {
-  source   = "../../..//modules/policy"
+module "policies" {
+  source   = "../../../modules/policy"
```
Comment (eemperor) [inline tests/create_roles/prereq/main.tf:6]: Same as previous comment.

File: tests/create_roles_policies/main.tf [hunk 6]
```diff
@@ -1,21 +1,9 @@
...
-  path           = each.value.path
-  template       = each.value.template
-  template_paths = each.value.template_paths
-  template_vars  = each.value.template_vars
-}
-
-module create_roles {
-  source   = "../..//modules/role"
+module "create_roles" {
+  source   = "../../modules/role"
```
Comment (eemperor) [inline tests/create_roles_policies/main.tf:6]: Same as previous comment.

File: tests/create_users/main.tf [hunk 7]
```diff
@@ -1,9 +1,9 @@
-provider aws {
+provider "aws" {
   region = "us-east-1"
 }
 
-module create_users {
-  source   = "../..//modules/user"
+module "create_users" {
+  source   = "../../modules/user"
```
Comment (eemperor) [inline tests/create_users/main.tf:6]: Same as previous comment.

File: tests/create_users/prereq/main.tf [hunk 8]
```diff
@@ -1,23 +1,56 @@
-provider aws {
+provider "aws" {
   region = "us-east-1"
 }
 
-module policies {
-  source   = "../../..//modules/policy"
+module "policies" {
+  source   = "../../../modules/policy"
```
Comment (eemperor) [inline tests/create_users/prereq/main.tf:6]: Same as previous comment.

File: tests/create_users_policies/main.tf [hunk 9]
```diff
@@ -1,21 +1,9 @@
...
-  path           = each.value.path
-  template       = each.value.template
-  template_paths = each.value.template_paths
-  template_vars  = each.value.template_vars
-}
-
-module create_users {
-  source   = "../..//modules/user"
+module "create_users" {
+  source   = "../../modules/user"
```
Comment (eemperor) [inline tests/create_users_policies/main.tf:6]: Same as previous comment.

**Review by eemperor — COMMENTED** [review]
Besides the `create_all` test, all the other tests are directly accessing the submodules.  Could they use the top-level module directly as well for consistency?

**Comment by lorengordon** [comment]
> Besides the create_all test, all the other tests are directly accessing the submodules. Could they use the top-level module directly as well for consistency?

No, the other tests are all intended to exercise the submodules independently. It is still valid for someone to call the submodules directly, if they want, so we also test that they work as standalone modules.

**Review by eemperor — APPROVED** [review]
Cool.  Looks good to me.
